pkg/ds/daemon_set.go

```diff
@@ -0,0 +1,353 @@
+	"github.com/square/p2/pkg/logging"
+	"github.com/square/p2/pkg/pods"
+	"github.com/square/p2/pkg/scheduler"
+	"github.com/square/p2/pkg/types"
+)
+
+const (
+	// This label is applied to pods owned by an RC.
+	DSIDLabel = "daemon_set_id"
+)
+
+type DaemonSet interface {
+	ID() fields.ID
+
+	WatchDesires(
+		quitCh <-chan struct{},
+		updatedCh <-chan fields.DaemonSet,
+		deletedCh <-chan struct{},
+		nodesChangedCh <-chan struct{},
+	) <-chan error
+
+	// CurrentNodes() returns all nodes that are scheduled by this daemon set
+	CurrentNodes() ([]types.NodeName, error)
+}
+
+type PodLocation struct {
+	Node  types.NodeName
+	PodID types.PodID
+}
+type PodLocations []PodLocation
+
+// Nodes returns a list of just the locations' nodes.
+func (l PodLocations) Nodes() []types.NodeName {
+	nodes := make([]types.NodeName, len(l))
+	for i, pod := range l {
+		nodes[i] = pod.Node
+	}
+	return nodes
+}
+
+// These methods are the same as the methods of the same name in kp.Store.
+// This interface makes writing unit tests easier as daemon sets do not
+// use any functions not listed here
+type kpStore interface {
+	SetPod(
+		podPrefix kp.PodPrefix,
+		nodeName types.NodeName,
+		manifest pods.Manifest,
+	) (time.Duration, error)
+
+	DeletePod(podPrefix kp.PodPrefix,
+		nodeName types.NodeName,
+		manifestID types.PodID,
+	) (time.Duration, error)
+}
+
+type daemonSet struct {
+	fields.DaemonSet
+
+	logger         logging.Logger
+	kpStore        kpStore
+	scheduler      scheduler.Scheduler
+	dsStore        dsstore.Store
+	nodeApplicator labels.Applicator
+}
+
+func New(
+	fields fields.DaemonSet,
+	dsStore dsstore.Store,
+	kpStore kpStore,
+	logger logging.Logger,
+	nodeApplicator labels.Applicator,
+	scheduler scheduler.Scheduler,
+) DaemonSet {
+	// TODO: Assert that arguments are not nil
+
+	return &daemonSet{
+		DaemonSet: fields,
+
+		dsStore:        dsStore,
+		kpStore:        kpStore,
+		logger:         logger,
+		nodeApplicator: nodeApplicator,
+		scheduler:      scheduler,
+	}
+}
+
+func (ds *daemonSet) ID() fields.ID {
+	return ds.DaemonSet.ID
+}
+
+// WatchDesires watches for changes to its daemon set, then schedule/unschedule
+// pods to to the nodes that it is responsible for
+//
+// Whatever calls WatchDesires is responsible for sending signals for whether
+// the daemon set updated or deleted
+//
+// When this is first called, it assumes that the daemon set is created
+//
+// The caller is responsible for sending signals when something has been changed
+func (ds *daemonSet) WatchDesires(
+	quitCh <-chan struct{},
+	updatedCh <-chan fields.DaemonSet,
+	deletedCh <-chan struct{},
+	nodesChangedCh <-chan struct{},
+) <-chan error {
+	errCh := make(chan error)
+
+	// Do something whenever something is changed
+	go func() {
+		var err error
+		defer close(errCh)
+
+		// Try to schedule pods when this begins watching
+		if !ds.Disabled {
+			ds.logger.NoFields().Infof("Received new daemon set: %v", *ds)
+			err = ds.addPods()
```

Comment:
`removePods()` will remove nodes(soon to be changed to pods) that have been labelled with the ds's uuid but not selected. Since nodes are first scheduled before the for loop at the beginning here, there shouldn't be any pods that have been labelled with the ds's uuid.